bdk_electrum_example/src/main.rs

```diff
@@ -0,0 +1,213 @@
+            }
+
+            // Fetch the new and old transactions to be added in update structure
+            for (tx, index) in fetch_transactions(&new_sparsechain, &client, &tracker)? {
+                keychain_scan.update.insert_tx(tx, index)?;
+            }
+
+            // Apply the full scan update
+            let changeset = tracker.determine_changeset(&keychain_scan)?;
+            db.append_changeset(&changeset)?;
+            tracker.apply_changeset(changeset);
+            debug!("sync completed!!")
+        }
+        ElectrumCommands::Sync {
+            mut unused,
+            mut unspent,
+            all,
+        } => {
+            let txout_index = &tracker.txout_index;
+            if !(all || unused || unspent) {
+                unused = true;
+                unspent = true;
+            } else if all {
+                unused = false;
+                unspent = false
+            }
+            let mut spks: Box<dyn Iterator<Item = bdk_core::bitcoin::Script>> =
+                Box::new(core::iter::empty());
+            if unused {
+                spks = Box::new(spks.chain(txout_index.iter_unused().map(|(index, script)| {
+                    eprintln!("Checking if address at {:?} has been used", index);
+                    script.clone()
+                })));
+            }
+
+            if all {
+                spks = Box::new(spks.chain(txout_index.script_pubkeys().iter().map(
+                    |(index, script)| {
+                        eprintln!("scanning {:?}", index);
+                        script.clone()
+                    },
+                )));
+            }
+
+            if unspent {
+                spks = Box::new(spks.chain(tracker.utxos().map(|(_index, ftxout)| {
+                    eprintln!("checking if {} has been spent", ftxout.outpoint);
+                    ftxout.txout.script_pubkey
+                })));
+            }
+
+            let mut scan_update = KeychainScan::default();
+
+            // Wallet scan returns a sparse chain that contains new All the BlockIds and Txids
+            // relevant to the wallet, along with keychain index update if required.
+            let new_sparsechain = client
+                .spk_txid_scan(spks, tracker.chain().checkpoints())
+                .context("scanning the blockchain")?;
+
+            // Inserting everything from the new_sparsechain should be okay as duplicate
+            // data would be rejected at the time of update application.
+            for (height, hash) in new_sparsechain.checkpoints() {
+                let _ = scan_update.update.insert_checkpoint(BlockId {
+                    height: *height,
+                    hash: *hash,
+                })?;
+            }
+
+            // Fetch the new and old transactions to be added in update structure
+            for (tx, index) in fetch_transactions(&new_sparsechain, &client, &tracker)? {
+                scan_update.update.insert_tx(tx, index)?;
+            }
```

Comment:
Can you try simplifying the above with the new APIs we added in https://github.com/LLFourn/bdk_core_staging/pull/92?

You can use `SpaseChain::changeset_additions` to figure out what the changeset is between your local sparse chain and the new sparse chain.. Fetch those full transactions and then call `ChainGraph::inflate_changeset` to turn your `sparse_chain::ChangeSet` into a `ChainGraph::ChangeSet`. You can then set the `KeychainChangeSet`'s `chain_graph` field with this value and just apply that changeset.

Reply:
Thanks this looks great.